src/nunit.analyzers/WithinUsage/WithinUsageAnalyzer.cs

```diff
@@ -0,0 +1,122 @@
+
+                if (expectedType == null || expectedType.TypeKind == TypeKind.Error)
+                    return;
+
+                if (!IsTypeSupported(expectedType))
+                {
+                    var syntax = withinSuffix.Syntax is InvocationExpressionSyntax expressionSyntax &&
+                        expressionSyntax.Expression is MemberAccessExpressionSyntax memberAccessSyntax
+                        ? memberAccessSyntax.Name
+                        : withinSuffix.Syntax;
+
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        descriptor,
+                        syntax.GetLocation()));
+                }
+            }
+        }
+
+        private static bool IsTypeSupported(ITypeSymbol type, HashSet<ITypeSymbol>? checkedTypes = null)
+        {
+            // Protection against possible infinite recursion
+            // TODO Remove suppression when the below is released:
+            // https://github.com/dotnet/roslyn-analyzers/issues/4568
+#pragma warning disable RS1024 // Compare symbols correctly
+            checkedTypes ??= new HashSet<ITypeSymbol>(SymbolEqualityComparer.Default);
+#pragma warning restore RS1024 // Compare symbols correctly
+            if (!checkedTypes.Add(type))
+                return false;
+
+            // Allowed - numeric or date types
+            if (type.SpecialType == SpecialType.System_Decimal ||
+                type.SpecialType == SpecialType.System_Double ||
+                type.SpecialType == SpecialType.System_Single ||
+                type.SpecialType == SpecialType.System_Char ||
+                type.SpecialType == SpecialType.System_Byte ||
+                type.SpecialType == SpecialType.System_SByte ||
+                type.SpecialType == SpecialType.System_Int16 ||
+                type.SpecialType == SpecialType.System_Int32 ||
+                type.SpecialType == SpecialType.System_Int64 ||
+                type.SpecialType == SpecialType.System_UInt16 ||
+                type.SpecialType == SpecialType.System_UInt32 ||
+                type.SpecialType == SpecialType.System_UInt64 ||
+                type.SpecialType == SpecialType.System_DateTime)
+            {
+                return true;
+            }
+
+            if (!(type is INamedTypeSymbol namedType))
+                return false;
+
+            // Allowed - tuples
+            if (namedType.IsTupleType)
+                return namedType.TupleElements.All(e => IsTypeSupported(e.Type, checkedTypes));
+
+            var fullName = namedType.GetFullMetadataName();
+
+            if (fullName.StartsWith("System.Tuple`", StringComparison.Ordinal))
+                return namedType.TypeArguments.All(t => IsTypeSupported(t, checkedTypes));
+
+            if (fullName.StartsWith("System.TimeSpan", StringComparison.Ordinal))
+                return true;
+
+            if (fullName.StartsWith("System.DateTime", StringComparison.Ordinal))
+                return true;
```

Comment:
Are you trying to match System.DateTimeOffset as System.DateTime is already checked earlier. Now you also match System.DateTimeKind which is not supported. If you want to match System.DateTimeOffset use Equals

Reply:
Ah I see. Thank you for spotting this!